### PR TITLE
Framework to migrate CI/CD pipelines to Github Actions

### PR DESCRIPTION
**Pull request by lisushka**
This pull request migrates the Linux and macOS tests from Azure Pipelines to Github Actions, as well as the deploy steps from Travis CI.  It also adds a shell task for migrating the Windows tests.

### CODE REVIEW

**Comment by Zac-HD** [comment]
Thanks @lisushka, this looks awesome :heart_eyes: 

For other reviewers: you can see what runs look like at https://github.com/lisushka/hypothesis/actions

@DRMacIver - you'll need to enable Actions for this repo before it'll have any effect.

**Comment by lisushka** [comment]
I can't get rebasing working, so for the authors file, please credit me as Dawn E. Collett.  =)

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      fail-fast: false
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+        architecture: ${{ matrix.python-architecture }}
+    - name: Install dependencies
+      run: pip install --upgrade setuptools pip wheel
```
Comment (sobolevn) [inline .github/workflows/main.yml:59]: It is not required, but sometimes very useful to have cache in-place. So, the installation process sometimes is skipped completely or partially. 

In our case it speeds up the build significantly. 
Ref: https://github.com/dry-python/returns/blob/master/.github/workflows/test.yml#L32-L36

Comment (sobolevn) [reply .github/workflows/main.yml:59]: Here are the numbers.

With the cache:
Снимок экрана 2020-09-07 в 15 25 17
https://github.com/dry-python/returns/actions/runs/242907292

Without the cache:
Снимок экрана 2020-09-07 в 15 25 37
https://github.com/dry-python/returns/runs/1081149090?check_suite_focus=true

Comment (lisushka) [reply .github/workflows/main.yml:59]: I think there was a brief discussion about caching in place while I was working on porting the CI/CD over, and Zac mentioned that there had been some issues with it on Azure Pipelines.  @Zac-HD Do I remember correctly?  If not, it should be trivial to give caching in place a try.

Comment (Zac-HD) [reply .github/workflows/main.yml:59]: We've had mixed experiences with caching in CI - in some cases it made our builds much slower :confused: 

So I think it's probably best to get things working first, and then we can look at speeding them up independently of the migration.

**Comment by lisushka** [comment]
Closing, as this has been superseded by#2674. 🙂

**Comment by Zac-HD** [comment]
Thanks again @lisushka - it's mostly your work and I couldn't have done it without you 😊

**Pull request closed without merging**